### ISSUE TITLE
`Outdated Optimize Dep` error when consuming 'normal' (direct) npm packages with vite – deps not ending up in `node_modules/.vite/deps`

### ISSUE THREAD

**Issue by johanrd** [embroider-build/embroider#2168]
## Steps to reproduce
1) Install the npm library `sanitize-html`, or clone this repositiory for reproduction: https://github.com/johanrd/ember-vite-npm

<details>
<summary>Detailed steps</summary>

1.1) create a new ember vite app with @NullVoxPopuli's great little ember-vite-blueprint `npx ember-cli@latest new glimmer-apollo-vite --blueprint @nullvoxpopuli/ember-vite-app --pnpm --typescript`

1.2) pnpm install any npm library, e.g. 'pnpm install sanitize-html`

1.3) import the pnpm library into wherever, e.g. application.gts

1.4) run `pnpm start`
</details>

2) See that the console errors with `Failed to load resource: the server responded with a status of 504 (Outdated Optimize Dep)`

<img width="1020" alt="Screenshot 2024-11-12 at 09 07 09" src="https://github.com/user-attachments/assets/bccb6602-bdba-4bd7-b07d-6ba0eeca1255">

Npm packages most often do not end up in `node_modules/.vite/deps`. Possibly related to https://github.com/embroider-build/embroider/issues/1583. 

Is there a config I am missing? I have tried to add `include` in optimizeDeps(), but that stopped the vite server completely.

**Comment by patricklx** [comment]
can you try running the command `vite optimize`? it is failing when I test it on your repo. But maybe that is related to my node_modules.

**Comment by patricklx** [comment]
right, thats what i am seeing as well `let terminalHighlight = require('./terminal-highlight');`
but why is it importing postcss into the app itself?

**Comment by patricklx** [comment]
oh, its used by sanitize-html.
just tried it on a repo without embroider and there the optimization of sanitize-html  works

**Comment by mansona** [comment]
I've had to fix this with sanatize-html before too. Unfortunately this isn't an embroider problem, I imagine you would have the exact same issue if you used it in a vanilla vite package too

I've fixed this in the past by patching postcss to add the extensions that it expects to be able to resolve that file: https://github.com/ember-learn/ember-api-docs/pull/920/files#diff-218116475a824ffaa3d495332d9bcf7dcd105c41f131ed71626cf1e7a66ac7d8

**Comment by patricklx** [comment]
@mansona it works in a vanilla vite app.
and it works if i add `if (path.includes('terminal-')) { return null; }` to all `build.onResolve` in `esbuild-resolver.js`

**Comment by patricklx** [comment]
I think embroider needs to handle the browser field: 

it is marked as excluded here:
https://github.com/postcss/postcss/blob/main/package.json#L132


https://github.com/defunctzombie/package-browser-field-spec?tab=readme-ov-file#ignore-a-module

**Comment by patricklx** [comment]
this is what i see in the deps of a vanilla vite repo:
![image](https://github.com/user-attachments/assets/b1536eb8-6dca-443e-87fa-88159c8201a1)

**Comment by patricklx** [comment]
some parts are related to this code:
https://github.com/embroider-build/embroider/blob/main/packages/vite/src/esbuild-request.ts#L199

it would also handle build-ins like `path`, but those are handled by vite with a custom namespace

**Comment by ef4** [comment]
I created a minimal reproduction of this bug in esbuild: https://github.com/evanw/esbuild/issues/3976

The summary is, esbuild knows to consider certain paths "disabled" by the `browser` entry in package.json, but it hides this fact from plugins. There's no way to round-trip the PathDisabled flag through a plugin in a way that esbuild itself will respect.

**Comment by patricklx** [comment]
@ef4 can we detect this in embroider and return null for these cases and have esbuild resolver handle it

**Comment by johanrd** [comment]
For the record, adding the context of https://discord.com/channels/480462759797063690/1355811390450958356/1355820755547586822 where the same thing happened with 'object-inspect'.

**Comment by johanrd** [comment]
Closing, as this is not reproducible with `vite@8.0.0-beta.2` rolldown
